src/hotspot/os/posix/perfMemory_posix.cpp

```diff
@@ -727,48 +715,97 @@ static void cleanup_sharedmem_resources(const char* dirname) {
-  // so, remove the file resources. Note, instrumented HotSpot processes
-  // for this user may start and/or terminate during this search and
-  // remove or create new files in this directory. The behavior of this
-  // loop under these conditions is dependent upon the implementation of
-  // opendir/readdir.
+  // For each entry in the directory that matches the expected file
+  // name pattern, remove the file if it's determine to be stale
+  // Note, instrumented HotSpot processes for this user may start and/or
+  // terminate during this search and remove or create new files in this
+  // directory. The behavior of this loop under these conditions is dependent
+  // upon the implementation of opendir/readdir.
   //
   struct dirent* entry;
   errno = 0;
   while ((entry = os::readdir(dirp)) != NULL) {
-
-    pid_t pid = filename_to_pid(entry->d_name);
+    const char* filename = entry->d_name;
+    pid_t pid = filename_to_pid(filename);
 
     if (pid == 0) {
 
-      if (strcmp(entry->d_name, ".") != 0 && strcmp(entry->d_name, "..") != 0) {
+      if (strcmp(filename, ".") != 0 && strcmp(filename, "..") != 0) {
         // attempt to remove all unexpected files, except "." and ".."
-        unlink(entry->d_name);
+        unlink(filename);
       }
 
       errno = 0;
       continue;
     }
 
-    // we now have a file name that converts to a valid integer
-    // that could represent a process id . if this process id
-    // matches the current process id or the process is not running,
-    // then remove the stale file resources.
+    // We now have a file name that converts to a valid integer
+    // that could represent a process id.
+    //
+    // On Linux, we first try to flock() on the file to check
+    // for JVM processes in other containers that share the same
+    // /tmp directory as the current process. If it's already
+    // locked by another process, then obviously it's not stale.
+    //
+    // Otherwise, if this process id matches the current process id or
+    // the process is not running, we treat it as a stale file and remove it.
     //
-    // process liveness is detected by sending signal number 0 to
-    // the process id (see kill(2)). if kill determines that the
-    // process does not exist, then the file resources are removed.
-    // if kill determines that that we don't have permission to
-    // signal the process, then the file resources are assumed to
-    // be stale and are removed because the resources for such a
+    // Process liveness is detected by sending signal number 0 to
+    // the process id (see kill(2)). If kill determines that the
+    // process does not exist, then the file is removed.
+    // If kill determines that that we don't have permission to
+    // signal the process, then the file is assumed to
+    // be stale and is removed because the files for such a
     // process should be in a different user specific directory.
```

Comment:
Please add a short sentence like:

"Finally, if VMs from different containers share /tmp, we cannot reliably determine process liveness using kill. To avoid that, either stop sharing /tmp between containers or only run VMs with the JDK-8286030 fix."